platform/linuxbsd/wayland/wayland_thread.cpp

```diff
@@ -4400,6 +4428,44 @@ void WaylandThread::pointer_set_hint(const Point2i &p_hint) {
 	}
 }
 
+void WaylandThread::pointer_warp(const Point2i &p_to) {
+	// NOTE: This is for compositors that don't support the pointer-warp protocol.
+	// It's hacked together and not guaranteed to work.
+	if (registry.wp_pointer_warp == nullptr) {
+		PointerConstraint old_constraint = pointer_get_constraint();
+
+		pointer_set_constraint(PointerConstraint::LOCKED);
+		pointer_set_hint(p_to);
+
+		pointer_set_constraint(old_constraint);
```

Comment:
random note: I think that the reason that this does not work is the fact that we do not wait for a frame/commit before setting the old constraint, as [the hint is double-buffered](https://wayland.app/protocols/pointer-constraints-unstable-v1#zwp_locked_pointer_v1:request:set_cursor_position_hint):

> The cursor position hint is double-buffered state, see wl_surface.commit.

We "could" do a roundtrip/commit there but committing in the middle of the frame is a bad, bad idea: #93684 (IIRC main thread commits are safe but the less the better)

IIRC this works on sway, which I think has a way more liberal interpretation, that's why I never noticed it.

To be clear, this is absolutely not something you should look in in this PR, it's for reference.

I really appreciate your PR, as there are lots of tiny rough edges here and there that I should look into (including good documentation about the backend) but my limited time is completely absorbed by the embedder PR, thank you.

Reply:
So i'll keep it as is?

Reply:
yes, absolutely. It was just a random note for reference.

Reply:
> I think that the reason that this does not work is the fact that we do not wait for a frame/commit before setting the old constraint

This seems to be the case, this version works extremely unreliable as is (one of 20 calls actually do anything), but with wait added in between constraint sets it seems to be perfectly working every time:

```suggestion
		wait_frame_suspend_ms(100);

		pointer_set_constraint(old_constraint);
```

Reply:
Thank you for looking into it! Your patch makes a lot of sense and IMO it's _almost_ perfect as is :D

I say almost because there are lots of small timing quirks to ensure, ideally.

Technically, what we're actually looking for is a _commit_, which applies the position hint, not a frame event per se. Frame events and commits are related as to request a frame callback you (or the renderer, or whatever) need to commit the surface but it could already have been requested by the time we receive it.

Also, the current logic basically assumes that surfaces are always committed from the _main thread_ every `DisplayServer::process_events` loop, either from the renderer or our own refresh logic for low consumption mode, which also gives some synchronization (mutex) concerns.

Probably the _ideal_ solution would be to have a callback of sorts at the start of the event loop to signal (somewhat late) to the Wayland thread that the previous main thread loop ended and only then reset the constraint. This is terribly not ideal, as technically there's a frame where the cursor is locked, but that's what the pointer constraints protocol entails FWIU.

I'll note this down somewhere and take a look whenever the time comes.